contracts/XINV.sol

```diff
@@ -0,0 +1,1087 @@
+    }
+
+    function getChainId() internal pure returns (uint) {
+        uint256 chainId;
+        assembly { chainId := chainid() }
+        return chainId;
+    }
+
+    function safe32(uint n, string memory errorMessage) internal pure returns (uint32) {
+        require(n < 2**32, errorMessage);
+        return uint32(n);
+    }
+
+    function add96(uint96 a, uint96 b, string memory errorMessage) internal pure returns (uint96) {
+        uint96 c = a + b;
+        require(c >= a, errorMessage);
+        return c;
+    }
+
+    function sub96(uint96 a, uint96 b, string memory errorMessage) internal pure returns (uint96) {
+        require(b <= a, errorMessage);
+        return a - b;
+    }
+}
+
+contract TimelockEscrow {
+    using SafeMath for uint;
+
+    address public underlying;
+    address public governance;
+    address public market;
+    uint public duration = 14 days;
+    mapping (address => EscrowData) public pendingWithdrawals;
+
+    struct EscrowData {
+        uint withdrawalTimestamp;
+        uint amount;
+    }
+
+    constructor(address underlying_, address governance_) public {
+        underlying = underlying_;
+        governance = governance_;
+        market = msg.sender;
+    }
+
+    // set to 0 to send funds directly to users
+    function _setEscrowDuration(uint duration_) public {
+        require(msg.sender == governance, "only governance can set escrow duration");
+        duration = duration_;
+    }
+
+    function _setGov(address governance_) public {
+        require(msg.sender == governance, "only governance can set its new address");
+        governance = governance_;
+    }
+
+    /**
+     * @notice assumes funds were already sent to this contract by the market. Resets escrow timelock on each withdrawal
+     */
+    function escrow(address user, uint amount) public {
+        require(msg.sender == market, "only market can escrow");
+        if(duration > 0) {
+            EscrowData memory withdrawal = pendingWithdrawals[user];
+            pendingWithdrawals[user] = EscrowData({
+                // we set the future withdrawal timestamp based on current `duration` to avoid applying future `duration` changes to existing withdrawals in the event of a governance attack
+                withdrawalTimestamp: block.timestamp + duration,
+                amount: withdrawal.amount.add(amount)
+            });
+            emit Escrow(user, block.timestamp + duration, amount);
+        } else { // if duration is 0, we send the funds directly to the user
+            EIP20Interface token = EIP20Interface(underlying);
+            token.transfer(user, amount);
+        }
+    }
+
+    /**
+     * @notice returns user withdrawable amount
+     */
+    function withdrawable(address user) public view returns (uint amount) {
+        EscrowData memory withdrawal = pendingWithdrawals[user];
+        if(withdrawal.withdrawalTimestamp > block.timestamp) {
```

Comment:
what happens if user doesn't withdraw at withdrawalTimestamp > block.timestamp . and then block.timestamp surpasses withdrawalTimestamp?
Is there a way to retrieve user funds in this case?

Reply:
withdraw() would then not transfer funds even though amount is actually gt 0, since withdrawable returns 0